osquery/tables/system/linux/services.cpp

```diff
@@ -0,0 +1,585 @@
+
+      for (const auto& run_level : run_level_list) {
+        if (run_level != 7) {
+          r["start_type"] += std::to_string(run_level);
+        } else {
+          r["start_type"] += "S";
+        }
+      }
+
+      r["start_type"] += "]";
+    }
+
+    query_data.push_back(r);
+  }
+
+  return Status(0, "OK");
+}
+
+sd_bus* getSystemdBusHandle() {
+  /*
+    https://github.com/systemd/systemd/blob/master/ENVIRONMENT.md
+
+    $SYSTEMCTL_FORCE_BUS=1
+      if set, do not connect to PID1's private D-Bus listener, and
+      instead always connect through the dbus-daemon D-bus broker.
+  */
+
+  auto force_dbus_env_var = std::getenv("SYSTEMCTL_FORCE_BUS");
+
+  bool connection_mode = true;
+  if (force_dbus_env_var != nullptr) {
+    connection_mode = boost::lexical_cast<bool>(force_dbus_env_var);
+  }
+
+  if (connection_mode || geteuid() != 0) {
+    sd_bus* bus = nullptr;
+    if (sd_bus_default_system(&bus) < 0) {
+      return nullptr;
+    }
+
+    sd_bus_set_allow_interactive_authorization(bus, 0);
+    return bus;
+  }
+
+  sd_bus* bus = nullptr;
+
+  try {
+    if (sd_bus_new(&bus) < 0 || bus == nullptr) {
+      throw std::runtime_error("Failed to start the bus");
+    }
+
+    if (sd_bus_set_address(bus, "unix:path=/run/systemd/private") < 0) {
+      throw std::runtime_error("Failed to set the bus address");
+    }
+
+    if (sd_bus_start(bus) < 0) {
+      throw std::runtime_error("Failed to start the bus");
+    }
+
+    sd_bus_set_allow_interactive_authorization(bus, 0);
+
+    auto fd = sd_bus_get_fd(bus);
+    if (fd < 0) {
+      throw std::runtime_error("Failed to acquire the bus file descriptor");
+    }
+
+    socklen_t l = sizeof(struct ucred);
+    struct ucred ucred = {};
+
+    if (getsockopt(fd, SOL_SOCKET, SO_PEERCRED, &ucred, &l) < 0) {
+      throw std::runtime_error("");
+    }
+
+    if (l != sizeof(struct ucred)) {
+      throw std::runtime_error("");
+    }
+
+    if (ucred.uid != 0 && ucred.uid != geteuid()) {
+      throw std::runtime_error("");
+    }
+
+    return bus;
+
+  } catch (const std::exception& e) {
+    VLOG(1) << e.what();
+
+    if (bus != nullptr) {
+      bus = sd_bus_flush_close_unref(bus);
+    }
+
+    return nullptr;
+  }
+}
+
+Status getSystemdUnitPid(pid_t& process_id,
+                         const std::string& unit_path,
+                         sd_bus* bus) {
+  sd_bus_message* reply = nullptr;
+  sd_bus_error bus_error = SD_BUS_ERROR_NULL;
+
+  try {
+    sd_bus_call_method(bus,
+                       "org.freedesktop.systemd1",
+                       unit_path.data(),
+                       "org.freedesktop.DBus.Properties",
+                       "GetAll",
+                       &bus_error,
+                       &reply,
+                       "s",
+                       "");
+
+    if (sd_bus_message_enter_container(reply, SD_BUS_TYPE_ARRAY, "{sv}") < 0) {
+      throw std::runtime_error("Failed to enter the data container");
+    }
+
+    while (true) {
```

Comment:
It is possible this can go infinitely? Could this implement a max-iterations to protect the execution?

Reply:
I don't think this can happen, as we are consuming bytes we have received from systemd using `sd_bus_call_method` (so they are limited).

How would you limit this? It seems like some services can have many properties (~241). We can maybe use a for loop from 1 to 1000 (to be safe). What do you think?